### PR TITLE
PRIME-1297 - PHSA Labtechs Backend

### PR DESCRIPTION
**Pull request by james-hollinger**
Backend for PHSA Labtechs eForms.
The secret for the keycloak administration client will not be committed to the project.

### CODE REVIEW

**Comment by neophyte57** [comment]
Is there some documentation on how this should work, especially re: what is created in database?  I don't see anything in https://bcgovmoh.atlassian.net/browse/PRIME-1297   Thanks.

File: prime-dotnet-webapi/Controllers/LabtechsController.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            if(await _partyService.UserIdExistsAsync(User.GetPrimeUserId()))
+            {
+                ModelState.AddModelError("Labtech.UserId", "A Labtech already exists with this User Id.");
+                return BadRequest(ApiResponse.BadRequest(ModelState));
+            }
+
+            Party model = null; //TODOD: temp Labtech.From(labtech, User);
+            await _partyService.CreatePartyAsync(model);
+
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/LabtechsController.cs:53]: `labtech` variable seems properly populated but why are we passing in null for the `model` variable?  System.ArgumentNullException gets thrown

**Comment by neophyte57** [comment]
1. From http://localhost:4200/phsa, I logged in with BCSC and went through the screens to save my phone #, extension, and email address.  Data gets saved to Party table 
2. Then I log out.  Go to http://localhost:4200/site and login with same BCSC id.  When I come back to site registration I get 
image and 
image

**Comment by neophyte57** [comment]
I'm not sure if my environment is properly updated.  I thought so.  I heard that there are corresponding deployments in Open Shift for each PR?  Where are they?

**Comment by johnnyl6** [comment]
Got error when using one of the PRIMETXXX account. Seems the backend api entouered 401 error but frontend just showing information has been saved, and not stays on current screen. Should the error be toasted to the user?

```
[13:28:47 ERR] Could not retrieve the role phsa_eforms_labtech from Keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
[13:28:47 ERR] Could not retrieve user cc8f2da6-f816-4d5a-be13-eea66032ac46 from keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
[13:28:47 INF] HTTP POST /api/parties/labtechs responded 200 in 102.3913 ms
```

Capture

**Comment by james-hollinger** [comment]
> 
> 
> Got error when using one of the PRIMETXXX account. Seems the backend api entouered 401 error but frontend just showing information has been saved, and stays on current screen. Should the error be toasted to the user?
> 
> ```
> [13:28:47 ERR] Could not retrieve the role phsa_eforms_labtech from Keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
> [13:28:47 ERR] Could not retrieve user cc8f2da6-f816-4d5a-be13-eea66032ac46 from keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
> [13:28:47 INF] HTTP POST /api/parties/labtechs responded 200 in 102.3913 ms
> ```
> 
> Capture

Communicating with Keycloak requires a secret I didn't want to commit to the repo. It is present in our openshift deployments, but until we have a better solution for local secrets I can send you the secret.

**Comment by johnnyl6** [comment]
> > Got error when using one of the PRIMETXXX account. Seems the backend api entouered 401 error but frontend just showing information has been saved, and stays on current screen. Should the error be toasted to the user?
> > ```
> > [13:28:47 ERR] Could not retrieve the role phsa_eforms_labtech from Keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
> > [13:28:47 ERR] Could not retrieve user cc8f2da6-f816-4d5a-be13-eea66032ac46 from keycloak. Response message: {"error":"HTTP 401 Unauthorized"}
> > [13:28:47 INF] HTTP POST /api/parties/labtechs responded 200 in 102.3913 ms
> > ```
> > 
> > 
> > Capture
> 
> Communicating with Keycloak requires a secret I didn't want to commit to the repo. It is present in our openshift deployments, but until we have a better solution for local secrets I can send you the secret.

work well after providing the secret. All good!

**Comment by johnnyl6** [comment]
> 1. From http://localhost:4200/phsa, I logged in with BCSC and went through the screens to save my phone #, extension, and email address.  Data gets saved to Party table
> 2. Then I log out.  Go to http://localhost:4200/site and login with same BCSC id.  When I come back to site registration I get
>    image and
>    image

Try a clean build of both frontend & backend. I don't get any error by following your steps.

**Review by johnnyl6 — APPROVED** [review]
🎉🎉🎉

File: prime-dotnet-webapi/Controllers/LabtechsController.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        /// If successful, also updates Keycloak with additional user info and the Labtech role.
+        /// </summary>
+        [HttpPost(Name = nameof(CreateLabtech))]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<ActionResult> CreateLabtech(LabtechChangeModel labtech)
+        {
+            await _partyService.CreateOrUpdatePartyAsync(labtech, User);
+
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/LabtechsController.cs:43]: This seems to be working properly:  I had an existing `Party` record, signing authority for an organization, and that record was updated.  Can we have a comment that this is the design?  The only documentation I have that this is the design are a few chat messages in "The Tech Team".

**Review by neophyte57 — CHANGES_REQUESTED** [review]

**Comment by neophyte57** [comment]
In https://bcgovmoh.atlassian.net/browse/PRIME-1208, there's a comment from @itchison:  "Individual profile data is common between all registrations and enrolments. Update your cell phone number during eform registration and pnet enrolment will gain the benefit of that new information. This allows the user to update their phone number once for the health sector instead of having to submit/update their information for every system they are enrolled/registered into."  

After completing my PHSA "registration" and having a record in the `Party` table, I switch to http://localhost:4200/info for PharmaNet enrollment.  Not sure if this is expected or not but on the Demographics page, the phone number and email address is not auto-populated from `Party` table.  When I complete PharmaNet enrollment, the `Party` is not updated with the different phone number and email address although a `Enrollee` record is created.  

Is this expected?

File: prime-dotnet-webapi/Controllers/LabtechsController.cs (hunk 2 above)
Comment (james-hollinger) [reply prime-dotnet-webapi/Controllers/LabtechsController.cs:43]: Yes, this is the design

**Comment by james-hollinger** [comment]
> 
> 
> In https://bcgovmoh.atlassian.net/browse/PRIME-1208, there's a comment from @itchison: "Individual profile data is common between all registrations and enrolments. Update your cell phone number during eform registration and pnet enrolment will gain the benefit of that new information. This allows the user to update their phone number once for the health sector instead of having to submit/update their information for every system they are enrolled/registered into."
> 
> After completing my PHSA "registration" and having a record in the `Party` table, I switch to http://localhost:4200/info for PharmaNet enrollment. Not sure if this is expected or not but on the Demographics page, the phone number and email address is not auto-populated from `Party` table. When I complete PharmaNet enrollment, the `Party` is not updated with the different phone number and email address although a `Enrollee` record is created.
> 
> Is this expected?

Enrollees are not (yet) Parties. At some point in the future they will be, but currently you might have a record in the Enrollee table and one record in the Party table.

**Comment by itchison** [comment]
Priority for this week

File: prime-dotnet-webapi/Controllers/LabtechsController.cs (hunk 2 above)
Comment (neophyte57) [reply prime-dotnet-webapi/Controllers/LabtechsController.cs:43]: Is it worthwhile to put this explicitly as a comment in the source code?

**Review by neophyte57 — APPROVED** [review]

**Pull request merged**